### PR TITLE
Add support for Ubuntu 14.04LTS Trusty Tahir

### PR DESCRIPTION
**Pull request by Aethylred**
This should have pam set up the same as the package release, though the comments are omitted from the files.

Small change to README.md stating that the lines parameters expect ordered arrays of strings.

### CODE REVIEW

**Comment by Aethylred** [comment]
Bother duplicates #73 but includes tests.

File: README.md [hunk 1]
```diff
@@ -180,25 +181,25 @@ Content template of $pam_d_sshd_path. If undef, parameter is set based on the OS
 
 pam_auth_lines
 --------------
-Content for PAM auth. If undef, parameter is set based on the OS version.
+An ordered array of strings that define the content for PAM auth. If undef, parameter is set based on the OS version.
```
Comment (ghoneycutt) [inline README.md]: Thanks for clarifying here!

File: manifests/init.pp [hunk 2]
```diff
@@ -575,8 +575,78 @@
                 ]
               }
             }
+            '14.04': {
+              $default_pam_d_login_template = 'pam/login.ubuntu14.erb'
+              $default_pam_d_sshd_template  = 'pam/sshd.ubuntu14.erb'
+              $default_package_name         = 'libpam0g'
+
+              if $ensure_vas == 'present' {
```
Comment (ghoneycutt) [inline manifests/init.pp]: Do you use VAS/QAS? If not, how did you arrive at these values?

**Comment by ghoneycutt** [comment]
Awesome work!

File: manifests/init.pp (hunk 2 above)
Comment (Aethylred) [inline manifests/init.pp]: uh... copy and paste...

**Comment by ghoneycutt** [comment]
@Phil-Friderici You know of any Ubuntu 14.04 systems with VAS to test with?

File: manifests/init.pp (hunk 2 above)
Comment (ghoneycutt) [inline manifests/init.pp]: understood, that should be fine for VAS

File: manifests/init.pp [hunk 3]
```diff
@@ -575,8 +575,78 @@
...
+                  'session optional  pam_umask.so',
+                  'session required  pam_vas3.so create_homedir',
+                  'session requisite pam_vas3.so echo_return',
+                  'session required  pam_unix.so'
+                ]
+
+              } else {
+
+                $default_pam_auth_lines = [
+                  'auth  [success=1 default=ignore]  pam_unix.so nullok_secure',
```
Comment (ghoneycutt) [inline manifests/init.pp]: These values also seem to be cut/paste. They should be taken directly from the affected files such as `/etc/pam.d/common-*`. The idea being that if you apply this module to a stock system, other than some formatting issues, it should not change how PAM works by default.

**Comment by Phil-Friderici** [comment]
@ghoneycutt I'll try to find one that I can use for testing purposes.

File: templates/sshd.ubuntu14.erb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+session    optional     pam_keyinit.so force revoke
+@include common-session
+session    optional     pam_motd.so  motd=/run/motd.dynamic noupdate
+session    optional     pam_motd.so # [1]
+session    optional     pam_mail.so standard noenv # [1]
+session    required     pam_limits.so
+session    required     pam_env.so # [1]
+session    required     pam_env.so user_readenv=1 envfile=/etc/default/locale
+session [success=ok ignore=ignore module_unknown=ignore default=bad]        pam_selinux.so open
+@include common-password
```
Comment (Phil-Friderici) [inline templates/sshd.ubuntu14.erb]: add a linebreak here, PLEASE ;)

**Comment by Phil-Friderici** [comment]
A test on Ubuntu 14.04 with VAS ran very well :+1: 

_[pedantic mode on]_
I am not so much into PAM details, so I don't know if the small changes that I have found does matter at all.

```diff
root@test:/tmp# diff -u login.clean /etc/pam.d/login
--- login.clean 2015-01-30 09:59:10.822141627 +0100
+++ /etc/pam.d/login    2015-01-30 10:23:58.073630373 +0100
@@ -2,7 +2,6 @@
 auth     [success=ok new_authtok_reqd=ok ignore=ignore user_unknown=bad default=die] pam_securetty.so
 auth     requisite  pam_nologin.so
 session  [success=ok ignore=ignore module_unknown=ignore default=bad] pam_selinux.so close
-session  required   pam_env.so readenv=1
 session  required   pam_env.so readenv=1 envfile=/etc/default/locale
 @include common-auth
 auth     optional   pam_group.so


root@edeaclx0600:/tmp# diff -u ssh.clean /etc/pam.d/sshd
--- ssh.clean   2015-01-30 09:53:17.496892758 +0100
+++ /etc/pam.d/sshd     2015-01-30 10:23:57.365643939 +0100
@@ -6,7 +6,10 @@
 session    optional     pam_keyinit.so force revoke
 @include common-session
 session    optional     pam_motd.so  motd=/run/motd.dynamic noupdate
+session    optional     pam_motd.so # [1]
+session    optional     pam_mail.so standard noenv # [1]
 session    required     pam_limits.so
+session    required     pam_env.so # [1]
 session    required     pam_env.so user_readenv=1 envfile=/etc/default/locale
 session [success=ok ignore=ignore module_unknown=ignore default=bad]        pam_selinux.so open
-@include common-password
+@include common-password
\ No newline at end of file
```

login.clean & ssh.clean: extracted from the deb packages, removed all comments
_[pedantic mode off]_

**Comment by ghoneycutt** [comment]
@Phil-Friderici The .clean is how you get it stock from the OS and the changes were after applying this module, correct?

**Comment by Phil-Friderici** [comment]
@ghoneycutt right, *.clean is from stock (deb package here) with all comments removed. And yes again, the changes were applyed by this code.

**Comment by ghoneycutt** [comment]
@Aethylred any objections to using the contents of those files as they came stock?

**Comment by Aethylred** [comment]
No objections.

File: templates/sshd.ubuntu14.erb (hunk 4 above)
Comment (Aethylred) [inline templates/sshd.ubuntu14.erb]: requires a trailing line break?

**Comment by Aethylred** [comment]
I've rebased against master and have to fix a few things.

**Comment by ghoneycutt** [comment]
@Aethylred No worries. Thanks for working on this!

**Comment by Aethylred** [comment]
Meeting finished early and now it's fixed.

**Comment by Phil-Friderici** [comment]
@Aethylred strange, I also did ```grep -v '^# ' /path/to/file```

**Comment by Phil-Friderici** [comment]
sorry for the mess, I have send a PR with the stock versions (```grep -v '^# ' /path/to/file```) to @Aethylred to fix that

File: manifests/init.pp [hunk 5]
```diff
@@ -765,10 +837,10 @@
   validate_re($sshd_pam_access, $valid_pam_access_values,
     "pam::sshd_pam_access is <${sshd_pam_access}> and must be either 'required', 'requisite', 'sufficient', 'optional' or 'absent'.")
 
-  if $package_name == undef {
```
Comment (ghoneycutt) [inline manifests/init.pp]: Hmm.. this logic should not have changed.

**Comment by ghoneycutt** [comment]
Merged @Phil-Friderici code into @Aethylred 's PR and reverted the logic around setting the package name in PR #99

**Pull request closed without merging**